clippy_lints/src/doc/broken_link.rs

```diff
@@ -0,0 +1,172 @@
+use clippy_utils::diagnostics::span_lint;
+use rustc_ast::{AttrKind, AttrStyle, Attribute};
+use rustc_lint::LateContext;
+use rustc_span::{BytePos, Span};
+
+use super::DOC_BROKEN_LINK;
+
+pub fn check(cx: &LateContext<'_>, attrs: &[Attribute]) {
+    for broken_link in BrokenLinkLoader::collect_broken_links(attrs) {
+        let reason_msg = match broken_link.reason {
+            BrokenLinkReason::MultipleLines => "broken across multiple lines",
+        };
+
+        span_lint(
+            cx,
+            DOC_BROKEN_LINK,
+            broken_link.span,
+            format!("possible broken doc link: {reason_msg}"),
+        );
+    }
+}
+
+/// The reason why a link is considered broken.
+// NOTE: We don't check these other cases because
+// rustdoc itself will check and warn about it:
+// - When a link url is broken across multiple lines in the URL path part
+// - When a link tag is missing the close parenthesis character at the end.
+// - When a link has whitespace within the url link.
+enum BrokenLinkReason {
+    MultipleLines,
+}
+
+/// Broken link data.
+struct BrokenLink {
+    reason: BrokenLinkReason,
+    span: Span,
+}
+
+enum State {
+    ProcessingLinkText,
+    ProcessedLinkText,
+    ProcessingLinkUrl(UrlState),
+}
+
+enum UrlState {
+    Empty,
+    FilledEntireSingleLine,
+    FilledBrokenMultipleLines,
+}
+
+/// Scan AST attributes looking up in doc comments for broken links
+/// which rustdoc won't be able to properly create link tags later.
+struct BrokenLinkLoader {
+    /// List of detected broken links.
+    broken_links: Vec<BrokenLink>,
+
+    state: Option<State>,
+
+    /// Keep track of the span for the processing broken link.
+    active_span: Option<Span>,
+
+    /// Keep track where exactly the link definition has started in the code.
+    active_pos_start: u32,
+}
+
+impl BrokenLinkLoader {
+    /// Return broken links.
+    fn collect_broken_links(attrs: &[Attribute]) -> Vec<BrokenLink> {
+        let mut loader = BrokenLinkLoader {
+            broken_links: vec![],
+            state: None,
+            active_pos_start: 0,
+            active_span: None,
+        };
+        loader.scan_attrs(attrs);
+        loader.broken_links
+    }
+
+    fn scan_attrs(&mut self, attrs: &[Attribute]) {
+        for attr in attrs {
+            if let AttrKind::DocComment(_com_kind, sym) = attr.kind
+                && let AttrStyle::Outer = attr.style
+            {
+                self.scan_line(sym.as_str(), attr.span);
+            }
+        }
+    }
+
+    fn scan_line(&mut self, line: &str, attr_span: Span) {
+        // Note that we specifically need the char _byte_ indices here, not the positional indexes
+        // within the char array to deal with multi-byte characters properly. `char_indices` does
+        // exactly that. It provides an iterator over tuples of the form `(byte position, char)`.
+        let char_indices: Vec<_> = line.char_indices().collect();
```

Comment:
This should just iterate directly rather than collecting.

Reply:
done